Update sliceviewer manual testing docs (#33420)
mantidproject/mantid: dev-docs/source/Testing/SliceViewer/SliceViewer.rst

See here for a brief overview of the :ref:`sliceviewer`.

Basic Usage
-----------

.. _toolbar-checklist:

MatrixWorkspace
###############

Do the following tests with an EventWorkspace (e.g. ``CNCS_7860_event.nxs``) and a Workspace2D (e.g. ``MAR11060.raw``) from the `TrainingCourseData <https://download.mantidproject.org/>`_.

> 1. Load the workspace and open in sliceviewer

Reviewer: Just for the very first time, can we say:
```suggestion
1. Load the workspace. Right-click on the workspace and select Show Slice Viewer.
```

> 2. Confirm that for MatrixWorkspaces the :ref`Peak overlay<mantid:sliceviewer_roi>` and :ref`Nonorthogonal view<mantid:sliceviewer_nonortho>` buttons should be disabled (greyed out).

Author: I think you might need to build all the docs as they are links to the sliceviewer interface docs

3. Check the other toolbar buttons.

    a. Pan and Zoom (use mouse scroll and magnifying glass tool) and Home

        * As you zoom check the color scale changes if autoscale checked in checkbox below colorbar).

    b. Toggle grid lines on/off
    c. Enable line plots

        * Confirm the curves update with the cursor position correctly

>     d. Export cuts to workspaces in ADS using keys `x`, `y`, `c`

Author: I've made a note on the issue that the extent of the cut along non-spectrum axis e.g. TOF is incorrect - it seems to grab more than the axes limits.

    e. Disable line plots and Enable :ref:`ROI tool<mantid:sliceviewer_roi>`

    .. figure:: ../../../../docs/source/images/wb-sliceviewer51-roibutton.png
       :class: screenshot
       :width: 50%
       :align: center

        * The line plot button should be automatically enabled

    f. Draw, move and resize the rectangle

        * Move it off the axes (it should just clip itself to be contained within the axes).
        * Export the cuts with keys `x`, `y`, `c`
        * In addition the ROI can be exported by pressing `r`

            - This should produce another workspace with suffix `_roi`
            - Open it in sliceviewer and check the data and limits agree with the ROI drawn.

    g.  Disable the ROI tool

        * This should automatically disable the line plot tool and the axes will be removed.

4. Try saving the figure (with and without ROI/lineplots).
5. Test the colorbar and colorscale

    a. Change normalisation

        * The color limits should only change if autoscale is enabled.

    b. Change the scale type to e.g. Log

        * In Log scale pixels with 0 counts should appear white

>         * When you zoom to a region with no data it will set the color axis limits to (0,0) and force the scale ot be linear

Author: I can't reproduce this - what OS are you on?

>         * When you zoom to a region with no data it will set the color axis limits to (0,0) and force the scale ot be linear

Reviewer: Ubuntu 18.04

    md_4D += tmp
    DeleteWorkspace(tmp)

    # Add a non-orthogonal UB
    expt_info = CreateSampleWorkspace()
    md_4D.addExperimentInfo(expt_info)
    SetUB(Workspace='md_4D', c=2, gamma=120)

    # make a 3D MDEvent workspace by integrating over all E
    md_3D = SliceMD(InputWorkspace='md_4D', AlignedDim0='H,-2,2,100', AlignedDim1='K,-1,1,100', AlignedDim2='L,-1.5,1.5,100')

    # Create a peaks workspace and fake data in 3D MD
    CreatePeaksWorkspace(InstrumentWorkspace='md_3D', NumberOfPeaks=0, OutputWorkspace='peaks')
    CopySample(InputWorkspace='md_3D', OutputWorkspace='peaks', CopyName=False, CopyMaterial=False, CopyEnvironment=False, CopyShape=False)
    AddPeakHKL(Workspace='peaks', HKL='1,0,1')
    AddPeakHKL(Workspace='peaks', HKL='1,0,0')

2. Test the toolbar buttons pan, zoom, line plots, ROI as in step 3 of the :ref:`toolbar-checklist` instructions.

    - This workspace should be dynamically rebinned - i.e. the number of bins within the view limits along each axis should be preserved when zooming.

3. Change the number of bins along one of the viewing axes (easier to pick a small number e.g. 2)

> 4. Change the integration width along the non-viewed axes.

Reviewer: ```suggestion
4. Change the thickness of the integration width along the non-viewed axes.
```
The word thick brings you to the right part of the GUI.

    - Increasing the width should improve the stats on the uniform background and the color limit should increase (event counts are summed not averaged).

5. Change the slicepoint along one of the non-viewed axes

    a. Confirm the slider moves when the spinbox value is updated.
    b. Confirm moving the slider updates the spinbox.

> Test the :ref`Nonorthogonal view<mantid:sliceviewer_nonortho>`

Reviewer: ```suggestion
Test the :ref:`Nonorthogonal view<mantid:sliceviewer_nonortho>`
```

.. figure:: ../../../../docs/source/images/wb-sliceviewer51-nonorthobutton.png
   :class: screenshot
   :align: center

> 1. Click the nonorthogonal view button in the toolbar

Reviewer: I got this error on `md_3D`:

``` python
Traceback (most recent call last):
  File "/home/danielmurphy/mantid/qt/python/mantidqt/mantidqt/widgets/sliceviewer/view.py", line 425, in on_non_orthogonal_axes_toggle
    self.presenter.nonorthogonal_axes(state)
  File "/home/danielmurphy/mantid/qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenter.py", line 365, in nonorthogonal_axes
    self.show_all_data_requested()
  File "/home/danielmurphy/mantid/qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenter.py", line 239, in show_all_data_requested
    self.set_axes_limits(*limits)
  File "/home/danielmurphy/mantid/qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenter.py", line 256, in set_axes_limits
    self.data_limits_changed()
  File "/home/danielmurphy/mantid/qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenter.py", line 226, in data_limits_changed
    self.new_plot()  # automatically uses current display limits
  File "/home/danielmurphy/mantid/qt/python/mantidqt/mantidqt/widgets/sliceviewer/presenter.py", line 125, in new_plot_MDE
    dimension_indices=dimension_indices))
  File "/home/danielmurphy/mantid/qt/python/mantidqt/mantidqt/widgets/sliceviewer/view.py", line 296, in plot_MDH_nonorthogonal
    self.draw_plot()
  File "/home/danielmurphy/mantid/qt/python/mantidqt/mantidqt/widgets/sliceviewer/view.py", line 352, in draw_plot
    self.colorbar.update_clim()
  File "/home/danielmurphy/mantid/qt/python/mantidqt/mantidqt/widgets/colorbar/colorbar.py", line 264, in update_clim
    self._autoscale_clim()
  File "/home/danielmurphy/mantid/qt/python/mantidqt/mantidqt/widgets/colorbar/colorbar.py", line 298, in _autoscale_clim
    data = self.colorbar.mappable.get_array_clipped_to_bounds()
AttributeError: 'QuadMesh' object has no attribute 'get_array_clipped_to_bounds'
```

Reviewer: I get the same error if I change displayed axes while in Nonoortho view

Reviewer: I also get these QuadMesh errors with MDHistoworkspace

    - This should disable ROI and lineplot buttons in the toolbar
    - This should automatically turn on gridlines
    - When H and K are the viewing axes the gridlines should not be perpendicular to each other
    - The features in the data should align with the grid lines

2. Zoom and pan

    - Confirm the autoscaling of the colorbar works in non-orthogonal view

3. Change one of the viewing axes to be `L` (e.g. click `X` button next to L in top left of window)

    - Gridlines should now appear to be orthogonal

4. For ``md_4D`` only change one of the viewing axes to be `E` (e.g. click `Y` button next to `E` in top left of window)

    - Nonorthogonal view should be disabled (only enabled for momentum axes)
    - Line plots and ROI should be enabled
    - Change the viewing axis presently selected as `E` to be a momentum axis (e.g. `H`)

            - The nonorthogonal view should be automatically re-enabled.

> Test the :ref`Peak Overlay<mantid:sliceviewer_peaks_overlay>`

Reviewer: ```suggestion
Test the :ref:`Peak Overlay<mantid:sliceviewer_peaks_overlay>`
```

2. Check the `Overlay?` box next to ``peaks``

    - This should open a table (peak viewer) on the RHS of the sliceviewer window - it should have a two rows corresponding to peaks at HKL = (1,0,1) and (1,0,0).

3. Double click a row

    - It should change the slicepoint along the integrated momentum axis and zoom into the peak - e.g. in (X,Y) = (H,K) then the slicepoint along L will be set to 1 and there will be a cross at (0,0).
    - Note for ``md_4D`` the cross should be plotted at all E (obviously a Bragg peak will only be on the elastic line but the peak object has no elastic/inelastic logic and the sliceviewer only knows that `E` is not a momentum axis, it could be temperature etc.).

4. Click Add Peaks in the Peak Actions section at the top of the peak viewer
5. Click somewhere in the colorfill plot

    - Confirm a peak has been added to the table at the position you clicked

6. Click Remove Peaks
7. Click on the cross corresponding to the peak you just added

    - Confirm the correct row has been removed from the table
    - The cross should be removed from the plot

8. Repeat the above steps 1-7 in non-orthogonal view.

MDHistoWorkspace
~~~~~~~~~~~~~~~~

> 1. Make a 2D and 3D MDHistoWorkspaces

Reviewer: ```suggestion
1. Make a 3D MDHistoWorkspaces
```
The code block only makes a 3D

Author: Ah good point  - the 2D test is covered by the `md_4D_svrebinned` stuff

.. code-block:: python

    md_3D_histo = BinMD(InputWorkspace='md_4D', AlignedDim0='H,-2,2,100', AlignedDim1='K,-1,1,100', AlignedDim2='L,-1.5,1.5,100')

2. Open ``md_3D_histo`` in sliceviewer it should not support dynamic rebinning (can't change number of bins).
3. Test the toolbar buttons pan, zoom, line plots, ROI as in step 3 of the :ref:`toolbar-checklist` instructions.
4. Test changing/swapping viewing axes
5. Test the :ref`Nonorthogonal view<mantid:sliceviewer_nonortho>` as above
6. Open ``md_4D_svrebinned`` in sliceviewer (should be in the ADS after preceding tests).

    - It should support dynamic rebinning (i.e. will be able to change number of bins along each axis).

7. With ``md_4D_svrebinned`` open in the sliceviewer, delete ``md_4D`` in the ADS.

>     - It should close sliceviewer because the support for dynamic rebinning has changed

Reviewer: It didn't close, but dynamic rebinning seems fine... I think the rebinning seemed fine because it had produced an ``md_4d_svrebinned_sv_rebinned`` workspace. But it doesn't close, even if I open ``md_4D_svrebinned``, do not zoom and delete md_4D

Author: Ah yes that's related to a memory leak - we know about that but I can't find the issue...

Author: Have made a new issue for this #33434

    # IntegratePeaksMD will throw an error
    #   Error in execution of algorithm MaskBTP:...
    # This is because the simulated ws don't have a real instrument but the integration will be executed

2. Open ``md_3D`` in sliceviewer
3. Click the peak overlay button in the toolbar
4. Overlay ``peaks_int_ellip`` and ``peaks_int_sphere``
5. Click the first row in the first table

    - It should zoom to a peak.
    - There should be an ellipse and a circle drawn with dashed lines with different colors (the color should match the color of the workspace name in the peak viewer table).
    - There should be a transparent shell indicating the background for each peak.
    - The ellipse should be smaller than the circle.

6. Alter the slice point by moving the slider along the integrated dimension

    - The circle and ellipse should shrink
    - There should be no gap between the background shell and the dashed line.

7. Click on the second row on the second table.

    - It should zoom in on a different peak.
    - The ellipse and circle should be very similar (not quite same as the covariance matrix was evaluated numerically for randomly generated data).

> 8. Click the nonorthogonal view button

Reviewer: testing these next few lines was blocked by the QuagMesh error!

    - Confirm the Peak actions combo box is updated to only contain ``peaks_int_ellipse``

4. Delete ``peaks_int_ellipse`` from the ADS

    - The peak overlay should be turned off and the table hidden

3. ADS observer for workspace
#############################

With ``MD_3D`` open in sliceviewer

1. Rename ``MD_3D`` to e.g. ``MD_3Dim``

    - The workspace name in the  title of the sliceviewer window should have updated
    - Zoom to check dynamic rebinning still works

2. Take a note of the colorbar limits and execute this command in the ipython terminal

    .. code-block:: python

        mtd['md_3Dim'] *= 2

    - The colorbar max should be doubled.
    - Zoom to check dynamic rebinning still works

3. Clone the workspace for future tests

.. code-block:: python

    CloneWorkspace(InputWorkspace='MD_3Dim', OutputWorkspace='MD_3D')

4. Delete ``MD_3Dim`` in the ADS

    - The sliceviewer window should close

4. ADS observer for support for nonorthogonal view
##################################################

1. Open ``MD_3D`` in sliceviewer
2. Run ``ClearUB`` algorithm on ``MD_3D``

    - Sliceviewer window should close with message

>     ``Closing Sliceviewer as the underlying workspace was changed: The property supports_nonorthogonal_axes is different on the new workspace.``

Reviewer: In sections 3. and 4. of the Specific tests you refer to ``md_3D`` as ``MD_3D``. This is mostly not a problem apart from the CloneWorkspace code line.

    SetUB(ws, 1,1,2,90,90,120)
    BinMD(InputWorkspace=ws, AxisAligned=False,
        BasisVector0='[00L],r.l.u.,0,0,1',
        BasisVector1='[HH0],r.l.u.,1,1,0',
        BasisVector2='[-HH0],r.l.u.,-1,1,0',
        OutputExtents='-4,4,-4,4,-0.25,0.25',
        OutputBins='101,101,1', OutputWorkspace='ws_slice', NormalizeBasisVectors=False)

2. Open ``ws_slice`` in the sliceviewer.

    - The non-orthogonal view should be enabled (not greyed out).

3. Click the non-orthogonal view button

    - Rectangular gridlines should appear (as in this case 110 is orthogonal to 001).

7. Check non-orthogonal view is disabled for non-Q axes
#######################################################
Check that the non-orthogonal view is disabled for non-Q axes such as energy

1. Create a workspace with energy as the first axis.

.. code-block:: python

    ws_4D = CreateMDWorkspace(Dimensions=4, Extents=[-1, 1, -1, 1, -1, 1, -1, 1], Names="E,H,K,L",
                                  Frames='General Frame,HKL,HKL,HKL', Units='meV,r.l.u.,r.l.u.,r.l.u.')
    expt_info_4D = CreateSampleWorkspace()
    ws_4D.addExperimentInfo(expt_info_4D)
    SetUB(ws_4D, 1, 1, 2, 90, 90, 120)

> 2. Confirm that when the Energy axis is viewed (as X or Y) the non-orthogonal view is disabled.

Reviewer: ```suggestion
2. Open `` ws_4D`` in sliceviewer. Confirm that when the Energy axis is viewed (as X or Y) the non-orthogonal view is disabled.
```
From the code snippet it's obvious but testers probably have lots in the ADS at this point!
